diff -Naur openssh-7.4p1/config.h openssh-7.4p1-android/config.h
--- openssh-7.4p1/config.h	2017-02-27 08:23:50.033399071 -0500
+++ openssh-7.4p1-android/config.h	2017-02-27 01:21:37.791822137 -0500
@@ -477,7 +477,7 @@
 #define HAVE_FSFILCNT_T 1
 
 /* Define to 1 if you have the `fstatfs' function. */
-#define HAVE_FSTATFS 1
+// #define HAVE_FSTATFS 1
 
 /* Define to 1 if you have the `fstatvfs' function. */
 /* #undef HAVE_FSTATVFS */
@@ -1057,7 +1057,7 @@
 #define HAVE_SS_FAMILY_IN_SS 1
 
 /* Define to 1 if you have the `statfs' function. */
-#define HAVE_STATFS 1
+// #define HAVE_STATFS 1
 
 /* Define to 1 if you have the `statvfs' function. */
 /* #undef HAVE_STATVFS */
@@ -1657,7 +1657,7 @@
 /* #undef UNIXWARE_LONG_PASSWORDS */
 
 /* Specify default $PATH */
-#define USER_PATH "/usr/bin:/bin:/usr/sbin:/sbin:/system/bin"
+#define USER_PATH "/system/bin:/system/xbin"
 
 /* Define this if you want to use libkafs' AFS support */
 /* #undef USE_AFS */
diff -Naur openssh-7.4p1/misc.c openssh-7.4p1-android/misc.c
--- openssh-7.4p1/misc.c	2016-12-18 23:59:41.000000000 -0500
+++ openssh-7.4p1-android/misc.c	2017-02-27 01:22:20.731821420 -0500
@@ -208,7 +208,9 @@
 	struct passwd *copy = xcalloc(1, sizeof(*copy));
 
 	copy->pw_name = xstrdup(pw->pw_name);
+#ifdef HAVE_STRUCT_PASSWD_PW_PASSWD
 	copy->pw_passwd = xstrdup(pw->pw_passwd);
+#endif
 #ifdef HAVE_STRUCT_PASSWD_PW_GECOS
 	copy->pw_gecos = xstrdup(pw->pw_gecos);
 #endif
@@ -223,7 +225,9 @@
 #ifdef HAVE_STRUCT_PASSWD_PW_CLASS
 	copy->pw_class = xstrdup(pw->pw_class);
 #endif
+#ifdef HAVE_STRUCT_PASSWD_PW_DIR
 	copy->pw_dir = xstrdup(pw->pw_dir);
+#endif
 	copy->pw_shell = xstrdup(pw->pw_shell);
 	return copy;
 }
@@ -1252,20 +1256,3 @@
 	return 1;
 }
 
-/* returns 1 if process is already daemonized, 0 otherwise */
-int
-daemonized(void)
-{
-	int fd;
-
-	if ((fd = open(_PATH_TTY, O_RDONLY | O_NOCTTY)) >= 0) {
-		close(fd);
-		return 0;	/* have controlling terminal */
-	}
-	if (getppid() != 1)
-		return 0;	/* parent is not init */
-	if (getsid(0) != getpid())
-		return 0;	/* not session leader */
-	debug3("already daemonized");
-	return 1;
-}
diff -Naur openssh-7.4p1/openbsd-compat/bsd-statvfs.c openssh-7.4p1-android/openbsd-compat/bsd-statvfs.c
--- openssh-7.4p1/openbsd-compat/bsd-statvfs.c	2016-12-18 23:59:41.000000000 -0500
+++ openssh-7.4p1-android/openbsd-compat/bsd-statvfs.c	2017-02-26 23:08:45.606955278 -0500
@@ -25,22 +25,6 @@
 
 #include <errno.h>
 
-static void
-copy_statfs_to_statvfs(struct statvfs *to, struct statfs *from)
-{
-	to->f_bsize = from->f_bsize;
-	to->f_frsize = from->f_bsize;	/* no exact equivalent */
-	to->f_blocks = from->f_blocks;
-	to->f_bfree = from->f_bfree;
-	to->f_bavail = from->f_bavail;
-	to->f_files = from->f_files;
-	to->f_ffree = from->f_ffree;
-	to->f_favail = from->f_ffree;	/* no exact equivalent */
-	to->f_fsid = 0;			/* XXX fix me */
-	to->f_flag = from->f_flags;
-	to->f_namemax = MNAMELEN;
-}
-
 # ifndef HAVE_STATVFS
 int statvfs(const char *path, struct statvfs *buf)
 {
diff -Naur openssh-7.4p1/openbsd-compat/explicit_bzero.c openssh-7.4p1-android/openbsd-compat/explicit_bzero.c
--- openssh-7.4p1/openbsd-compat/explicit_bzero.c	2016-12-18 23:59:41.000000000 -0500
+++ openssh-7.4p1-android/openbsd-compat/explicit_bzero.c	2017-02-26 23:13:31.922950496 -0500
@@ -25,12 +25,6 @@
 
 #else /* HAVE_MEMSET_S */
 
-/*
- * Indirect bzero through a volatile pointer to hopefully avoid
- * dead-store optimisation eliminating the call.
- */
-static void (* volatile ssh_bzero)(void *, size_t) = bzero;
-
 void
 explicit_bzero(void *p, size_t n)
 {
@@ -45,7 +39,7 @@
 # endif
 #endif
 
-	ssh_bzero(p, n);
+	bzero(p, n);
 }
 
 #endif /* HAVE_MEMSET_S */
diff -Naur openssh-7.4p1/sftp-server.c openssh-7.4p1-android/sftp-server.c
--- openssh-7.4p1/sftp-server.c	2016-12-18 23:59:41.000000000 -0500
+++ openssh-7.4p1-android/sftp-server.c	2017-02-27 01:24:26.216819324 -0500
@@ -1554,7 +1554,13 @@
 			break;
 		case 'd':
 			cp = tilde_expand_filename(optarg, user_pw->pw_uid);
-			homedir = percent_expand(cp, "d", user_pw->pw_dir,
+			homedir = percent_expand(cp,
+          "d",
+#ifdef HAVE_STRUCT_PASSWD_PW_DIR
+          user_pw->pw_dir,
+#else
+          "/mnt/sdcard"
+#endif
 			    "u", user_pw->pw_name, (char *)NULL);
 			free(cp);
 			break;
